### PR TITLE
Password prompt choices

### PR DESCRIPTION
**Pull request by configurator**
This PR adds an `XSECURELOCK_PASSWORD_PROMPT` environment variable, which overrides `XSECURELOCK_PARANOID_PASSWORD` with more options.

* `cursor`: shows a cursor that jumps around on each key press.
* `asterisks`: shows asterisks, like classic password prompts. This is less secure because password length is visible.
* `hidden`: completely hides the password, and there's no feedback for keypresses.
* `disco`: shows dancers, which dance around on each key press.
* `emoji`: shows an emoji, changing which one on each key press. (Requires using an emoji-supporting font)
* `emoticon`: shows an ascii emoticon, changing which one on each key press.
* `kaomoji`: shows a kaomoji (Japanese emoticon), changing which one on each key press.

`XSECURELOCK_PARANOID_PASSWORD` is marked as deprecated in the readme, so that users would proper using `XSECURELOCK_PASSWORD_PROMPT=<choice>` instead.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -309,8 +309,21 @@ Options to XSecureLock can be passed by environment variables:
     and fall back to XRandR 1.2. Not recommended.
 *   `XSECURELOCK_PAM_SERVICE`: pam service name. You should have a file with
     that name in `/etc/pam.d`.
-*   `XSECURELOCK_PARANOID_PASSWORD`: make `auth_x11` hide the password
-    length.
+*   `XSECURELOCK_PARANOID_PASSWORD`: (deprecated) makes `auth_x11` hide the
```
Comment (divVerent) [inline README.md]: Let's actually remove it from the README.

Also, in the code, can you please add a comment like here https://github.com/google/xsecurelock/blob/6f8876ee0eef0556f15ac06568944cbb20663e1d/helpers/dimmer.c#L291 (yes, _inside_ the string so ./ensure-documented-settings.sh can no longer see it) so I can't forget to remove the old setting in v2.0.0?

File: README.md [hunk 2]
```diff
@@ -309,8 +309,21 @@ Options to XSecureLock can be passed by environment variables:
...
+    password length. Use `XSECURELOCK_PASSWORD_PROMPT` instead. (`1` translates
+    to `cursor`; `0` translates to asterisks)
+*   `XSECURELOCK_PASSWORD_PROMPT`: Choose password prompt mode:
+    *   `cursor`: shows a cursor that jumps around on each key press.
```
Comment (divVerent) [inline README.md]: Can you put example strings in the doc here?

File: helpers/auth_x11.c [hunk 3]
```diff
@@ -62,11 +63,14 @@ const char *authproto_executable;
...
+//! Number of dancers in the disco password display
+#define DISCO_PASSWORD_DANCERS 5
+
 //! Length of the "paranoid password display".
-#define PARANOID_PASSWORD_LENGTH 32
+#define PARANOID_PASSWORD_LENGTH (1 << DISCO_PASSWORD_DANCERS)
 
 //! Minimum distance the cursor shall move on keypress.
-#define PARANOID_PASSWORD_MIN_CHANGE 4
+#define PARANOID_PASSWORD_MIN_CHANGE 5
```
Comment (divVerent) [inline helpers/auth_x11.c]: Not objecting, but why?

File: helpers/auth_x11.c [hunk 4]
```diff
@@ -77,8 +81,86 @@ int prompt_timeout;
 //! Extra line spacing.
 #define LINE_SPACING 4
 
-//! Whether password display should hide the length.
-int paranoid_password;
+//! Deprecated flag for setting Whether password display should hide the length.
+int paranoid_password_flag;
```
Comment (divVerent) [inline helpers/auth_x11.c]: No longer needs to be global; please move these variables into main() now.

File: helpers/auth_x11.c [hunk 5]
```diff
@@ -77,8 +81,86 @@ int prompt_timeout;
...
+  "_____", "\U0001f602", "\U00002764", "\U0000267b",
+  "\U0001f60d", "\U00002665", "\U0001f62d", "\U0001f60a",
+  "\U0001f612", "\U0001f495", "\U0001f618", "\U0001f629",
+  "\U0000263a", "\U0001f44c", "\U0001f614", "\U0001f601",
+  "\U0001f60f", "\U0001f609", "\U0001f44d", "\U00002b05",
+  "\U0001f605", "\U0001f64f", "\U0001f60c", "\U0001f622",
+  "\U0001f440", "\U0001f494", "\U0001f60e", "\U0001f3b6",
+  "\U0001f499", "\U0001f49c", "\U0001f64c", "\U0001f633",
+};
+_Static_assert(sizeof(emojis) / sizeof(*emojis) == PARANOID_PASSWORD_LENGTH, "Emojis array size must be equal to PARANOID_PASSWORD_LENGTH");
```
Comment (divVerent) [inline helpers/auth_x11.c]: Given this is our only use of C11 anywhere, please put the _Static_asserts into an

```
#if __STDC_VERSION__ >= 201112L
```

block. Thanks!

File: helpers/auth_x11.c [hunk 6]
```diff
@@ -77,8 +81,86 @@ int prompt_timeout;
...
+const char * disco_dancers[] = {
+  "┏(･o･)┛",
+  "┗(･o･)┓",
+};
+
+// Emojis to display in emoji mode. The length of the array must be equal to PARANOID_PASSWORD_LENGTH.
+// List taken from the top items in http://emojitracker.com/
+// The first item is always display in an empty prompt (before typing in the password)
+const char * emojis[] = {
+  "_____", "\U0001f602", "\U00002764", "\U0000267b",
```
Comment (divVerent) [inline helpers/auth_x11.c]: You've put all the other Unicode characters in there directly - why not these? I.e. actually put the emoji in here as UTF-8, like in the others. More readable.

File: helpers/auth_x11.c [hunk 7]
```diff
@@ -936,30 +1024,77 @@ int Prompt(const char *msg, char **response, int echo) {
...
+            memcpy(priv.displaybuf + i * stride, disco_combiner, combiner_length);
+            memcpy(priv.displaybuf + i * stride + combiner_length, dancer, dancer_length);
+          }
+          memcpy(priv.displaybuf + DISCO_PASSWORD_DANCERS * stride, disco_combiner, combiner_length);
+          priv.displaybuf[priv.displaylen] = '\0';
+          break;
+        }
+
+        case PASSWORD_PROMPT_EMOJI: {
+          showFromArray(emojis, priv.displaymarker, &(priv.displaybuf), &priv.displaylen);
```
Comment (divVerent) [inline helpers/auth_x11.c]: Rename: ShowFromArray. Otherwise good.

File: helpers/auth_x11.c [hunk 8]
```diff
@@ -936,30 +1024,77 @@ int Prompt(const char *msg, char **response, int echo) {
...
+          break;
+        }
+
+        case PASSWORD_PROMPT_DISCO: {
+          size_t combiner_length = strlen(disco_combiner);
+          size_t dancer_length = strlen(disco_dancers[0]);
+          size_t stride = combiner_length + dancer_length;
+          priv.displaylen = stride * DISCO_PASSWORD_DANCERS * strlen(disco_dancers[0]) + strlen(disco_combiner);
+
+          for (size_t i = 0, bit = 1; i < DISCO_PASSWORD_DANCERS; i++, bit <<= 1) {
```
Comment (divVerent) [inline helpers/auth_x11.c]: nit: ++i (consistency)

File: helpers/auth_x11.c (hunk 7 above)
Comment (divVerent) [inline helpers/auth_x11.c]: nit: &priv.displaybuf

File: README.md (hunk 2 above)
Comment (configurator) [reply README.md]: I'm not sure what you mean by this comment. Could you expand?

File: README.md (hunk 1 above)
Comment (configurator) [reply README.md]: Fixed
`README.md` - 4ccc9d4
`auth_x11.c` - b1701ec

File: helpers/auth_x11.c (hunk 3 above)
Comment (configurator) [reply helpers/auth_x11.c]: Accidentally changed - reverted in 9e72fea.

File: helpers/auth_x11.c (hunk 5 above)
Comment (configurator) [reply helpers/auth_x11.c]: How about a macro?

```
#if __STDC_VERSION__ >= 201112L
#define ASSERT(state, message) _Static_assert(state, message)
#else
#define ASSERT(state, message)
#endif
```

Implemented in c1cfe35 for your review.

File: helpers/auth_x11.c (hunk 6 above)
Comment (configurator) [reply helpers/auth_x11.c]: Done in fecdb1b.

File: helpers/auth_x11.c (hunk 7 above)
Comment (configurator) [reply helpers/auth_x11.c]: Done in b12255d.

Comment (configurator) [reply helpers/auth_x11.c]: Done in eecbdf8.

File: helpers/auth_x11.c (hunk 8 above)
Comment (configurator) [reply helpers/auth_x11.c]: Done in 70bc5a0.

File: helpers/auth_x11.c (hunk 4 above)
Comment (configurator) [reply helpers/auth_x11.c]: Done in 45e0926.

File: README.md (hunk 2 above)
Comment (divVerent) [reply README.md]: Something like
`_________/______`
`********_`
etc.

File: helpers/auth_x11.c (hunk 5 above)
Comment (divVerent) [reply helpers/auth_x11.c]: Yeah, that will do.

File: README.md (hunk 2 above)
Comment (configurator) [reply README.md]: Added in 833cf0d2fe46d71014207395d23153d2688ea49c.

Comment (configurator) [reply README.md]: (see https://github.com/configurator/xsecurelock/blob/prompts-pr/README.md#options)

**Pull request merged**